### DISCUSSION TITLE
psycopg.errors.IndeterminateDatatype. Fastest and painless solution / workaround to find all places to fix

### DISCUSSION THREAD

**Discussion by 91DarioDev** [psycopg/psycopg#333] [Q&A] [answered]
after i completed the migration of the library i am experiencing this error:
`psycopg.errors.IndeterminateDatatype: non è stato possibile determinare il tipo di dato del parametro $14`

in a portion of a query like this 

`AND CASE WHEN (%s IS NULL) THEN partner_min_age=%s AND partner_max_age=%s`

it happens in the first placeholder of this query and the values i'm passing is `None`. so that it will enter the else and other logic.

i searched on the documentation and i found it is explained [here](https://www.psycopg.org/psycopg3/docs/basic/from_pg2.html#different-cast-rules)

so if i understood it fine i have to explicity cast to the the type and since that becomes from a previous select query from a `DATE` column i guess and i should make this:

`AND CASE WHEN (%s::DATE IS NULL) THEN partner_min_age=%s AND partner_max_age=%s`

it seems it's working. so is it right? btw the most important part is: since i noticed it happens in many many queries, is there a way to find out where it can happen? or should i test all the queries in the project again and with all the possbile parameters a placeholder could have? what could be the painless solution??

thanks

**Comment by dvarrazzo** [comment, marked as answer]
Yes, your solution is right.

The problem is with queries where a parameter is in a context that gives no type indication whatsoever. Typically, a placeholder in an insert or in a comparison with a typed expression gives enough information to resolve an undefined type.

In the docs we mentioned variadic functions, which are not extremely common (`concat()`, for instance). I think the `IS NULL` case is actually a more common occurrence of this problem. It's too bad that actually the type of the argument is irrelevant...

If we solve the `IS [NOT] NULL` case which are probably simple to grep, do you find other types of problematic queries in your codebase?

**Comment by 91DarioDev** [comment]
Thank you very much. It seems in my case I solved pretty fast searching in all the project 

`%s IS`
